erpnext/hr/doctype/salary_structure/salary_structure.py

```diff
@@ -29,12 +29,34 @@ def validate_amount(self):
 		if flt(self.net_pay) < 0 and self.salary_slip_based_on_timesheet:
 			frappe.throw(_("Net pay cannot be negative"))
 
-	def validate_joining_date(self):
-		for e in self.get('employees'):
-			joining_date = getdate(frappe.db.get_value("Employee", e.employee, "date_of_joining"))
-			if e.from_date and getdate(e.from_date) < joining_date:
+	def validate_date(self):
+		for employee in self.get('employees'):
+			joining_date, relieving_date = frappe.db.get_value("Employee", employee.employee,
+				["date_of_joining", "relieving_date"])
+			if employee.from_date and getdate(employee.from_date) < joining_date:
 				frappe.throw(_("From Date {0} for Employee {1} cannot be before employee's joining Date {2}")
-					    .format(e.from_date, e.employee, joining_date))	
+					    .format(employee.from_date, employee.employee, joining_date))
+
+		st_name = frappe.db.sql("""select parent from `tabSalary Structure Employee`
+			where
+			employee=%(employee)s
+			and (
+				(%(from_date)s between from_date and ifnull(to_date, '2199-12-31'))
+				or (%(to_date)s between from_date and ifnull(to_date, '2199-12-31'))
+				or (from_date between %(from_date)s and ifnull(%(to_date)s, '2199-12-31'))
+			)
+			and parent in (select name from `tabSalary Structure`
```

Comment:
"in" query becomes very slow if there are too many values in sub query.

Instead you can write like this:
`and exists(select name from `tabSalary Structure` where name=`tabSalary Structure Employee`.parent and is_active = 'Yes') and parent != %(salary_struct)s`